File: apps/embedding-worker/src/logger.py
"""Structured JSON logger for the embedding worker.

Mirrors the TypeScript @hybrid-memory/observability logger contract:
  { timestamp, level, service, component, event, trace_id?, tenant_id?, ... }

Usage:
    from src.logger import create_logger

    log = create_logger("consumer")
    log.info("job_received", memory_id="abc-123", chunk_count=5)

    # Bind persistent context fields
    job_log = log.with_context(memory_id="abc-123", trace_id="t-1")
    job_log.info("processing")
"""

import json
import sys
import time
from typing import Any

SERVICE = "embedding-worker"


class Logger:
    def __init__(self, component: str = "main", context: dict[str, Any] | None = None):
        self._component = component
        self._context: dict[str, Any] = context or {}

    def _emit(self, level: str, event: str, **kwargs: Any) -> None:
        entry: dict[str, Any] = {
            "timestamp": time.strftime("%Y-%m-%dT%H:%M:%S%z"),
            "level": level,
            "service": SERVICE,
            "component": self._component,
            "event": event,
        }
        # Merge bound context, then call-site kwargs (call-site wins)
        entry.update(self._context)
        entry.update({k: v for k, v in kwargs.items() if v is not None})

        line = json.dumps(entry)
        if level == "error":
            sys.stderr.write(line + "\n")
            sys.stderr.flush()
        else:
            sys.stdout.write(line + "\n")
            sys.stdout.flush()

    def debug(self, event: str, **kwargs: Any) -> None:
        self._emit("debug", event, **kwargs)

    def info(self, event: str, **kwargs: Any) -> None:
        self._emit("info", event, **kwargs)

    def warn(self, event: str, **kwargs: Any) -> None:
        self._emit("warn", event, **kwargs)

    def error(self, event: str, **kwargs: Any) -> None:
        self._emit("error", event, **kwargs)

    def child(self, component: str) -> "Logger":
        return Logger(component, dict(self._context))

    def with_context(self, **ctx: Any) -> "Logger":
        merged = {**self._context, **{k: v for k, v in ctx.items() if v is not None}}
        return Logger(self._component, merged)


def create_logger(component: str = "main") -> Logger:
    return Logger(component)

File: packages/observability/src/logger.ts
import { getTraceContext } from "./traceContext";

export type LogLevel = "debug" | "info" | "warn" | "error";

const LEVEL_ORDER: Record<LogLevel, number> = {
  debug: 0,
  info: 1,
  warn: 2,
  error: 3,
};

/** Standard context fields that can be bound to a logger instance. */
export interface LogContext {
  tenant_id?: string;
  workspace_id?: string;
  user_id?: string;
  memory_id?: string;
  fact_id?: string;
  job_id?: string;
  [key: string]: string | undefined;
}

export interface LogEntry {
  timestamp: string;
  level: LogLevel;
  service: string;
  component: string;
  event: string;
  trace_id?: string;
  tenant_id?: string;
  workspace_id?: string;
  user_id?: string;
  memory_id?: string;
  fact_id?: string;
  job_id?: string;
  duration_ms?: number;
  status?: string | number;
  [key: string]: unknown;
}

export interface Logger {
  debug(event: string, data?: Record<string, unknown>): void;
  info(event: string, data?: Record<string, unknown>): void;
  warn(event: string, data?: Record<string, unknown>): void;
  error(event: string, data?: Record<string, unknown>): void;
  child(component: string): Logger;
  /** Return a new Logger with the given context fields bound to every entry. */
  withContext(ctx: LogContext): Logger;
}

let minLevel: LogLevel = "info";

export function setLogLevel(level: LogLevel): void {
  minLevel = level;
}

export function createLogger(
  service: string,
  component = "main",
  boundContext: LogContext = {}
): Logger {
  function log(
    level: LogLevel,
    event: string,
    data?: Record<string, unknown>
  ): void {
    if (LEVEL_ORDER[level] < LEVEL_ORDER[minLevel]) return;

    const ctx = getTraceContext();

    // Merge: trace attributes < bound context < call-site data
    const entry: LogEntry = {
      timestamp: new Date().toISOString(),
      level,
      service,
      component,
      event,
      ...(ctx?.traceId ? { trace_id: ctx.traceId } : {}),
      ...stripUndefined(ctx?.attributes ?? {}),
      ...stripUndefined(boundContext),
      ...data,
    };

    const output = JSON.stringify(entry);
    if (level === "error") {
      process.stderr.write(output + "\n");
    } else {
      process.stdout.write(output + "\n");
    }
  }

  return {
    debug: (event, data) => log("debug", event, data),
    info: (event, data) => log("info", event, data),
    warn: (event, data) => log("warn", event, data),
    error: (event, data) => log("error", event, data),
    child: (childComponent) => createLogger(service, childComponent, boundContext),
    withContext: (ctx) => createLogger(service, component, { ...boundContext, ...ctx }),
  };
}

function stripUndefined(obj: Record<string, unknown>): Record<string, unknown> {
  const result: Record<string, unknown> = {};
  for (const [k, v] of Object.entries(obj)) {
    if (v !== undefined) result[k] = v;
  }
  return result;
}
